### PR TITLE
[Dev] Support EP with HSDP

### PR DESCRIPTION
**Pull request by wplf**
# What does this PR do ?
main PR: [#2840](https://github.com/NVIDIA/Megatron-LM/pull/2840) 

This MR adds HSDP support for Expert Parallelism. With this change, DeepSeek‑v3 and other MoE models can be trained using M‑FSDP as well as HSDP+EP. HSDP enables better use of local bandwidth within a larger nvlink domain.

We ran convergence tests on the [DeepSeek Proxy model](https://gitlab-master.nvidia.com/arch_moe_exploration/megatron-moe-scripts/-/blob/master/model_configs/benchmarking/DeepSeek-V3-Proxy.yaml?ref_type=heads). model. The experiments were based on the latest dev branch, with an additional [patch](https://github.com/xuwchen/Megatron-LM/commit/2d8a285a4964202567f178f1c0f2e6ab25aa1a54) that fixes MoE checkpoint save/load. 

To validate the correctness of HSDP+EP, we loaded an FSDP+EP checkpoint at step 100 and resumed training with HSDP+EP. The loss curves and gradient norms matched closely.

image
image

For example, to use HSDP=2 and EP size=2, you need to add these argument to your script 
```bash
--num-distributed-optimizer-instances 2 \
--outer-dp-sharding-strategy optim
--expert-model-parallel-size 2 
```

### CODE REVIEW

File: megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py [hunk 1]
```diff
@@ -2005,7 +2005,7 @@ def _init_each_parameter_group_buffers(self, meta_device_init_fp8_params):
         # Specifically, replace the Torch module's parameter data with tensors
         # whose memory managed by the model weight buffer, and store a shard
         # of all the parameters across ranks in the model weight buffer.
-        for group in self.parameter_groups:
+        for group_id, group in enumerate(self.parameter_groups):
```
Comment (lilei199908) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py]: why need group_id ? where to use?

File: megatron/core/optimizer/optimizer_config.py [hunk 2]
```diff
@@ -171,6 +171,9 @@ class OptimizerConfig:
     overlap_param_gather_with_optimizer_step: bool = False
     """If true, overlap param all-gather of first bucket with optimizer step."""
 
+    num_distributed_optimizer_instances: int = 1
```
Comment (BoxiangW) [inline megatron/core/optimizer/optimizer_config.py]: Why are we adding this config here? It was under ddp_config before

File: megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py (hunk 1 above)
Comment (wplf) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py]: You're right. I've deleted it.

File: megatron/core/optimizer/optimizer_config.py (hunk 2 above)
Comment (wplf) [reply megatron/core/optimizer/optimizer_config.py]: You're right. I've deleted it.

**Comment by wplf** [comment]
@BoxiangW
Hi, boxiang.
The loss convergency and grad norm curve of HSDP+EP vs FSDP+EP have been tested.
Is there anything else that I need to do?

File: megatron/core/distributed/fsdp/src/README.md [hunk 3]
```diff
@@ -144,13 +144,13 @@ device_mesh[("dp_shard", "cp")]._flatten("dp_shard_cp")
 # Only required if using HSDP. Otherwise, don't pass hybrid_fsdp_group.
 device_mesh[("dp_outer", "dp_shard", "cp")]._flatten("hsdp")
 hsdp_group = device_mesh["hsdp"].get_group()
-# Initialize DeviceMesh for expert parallel (EP) modules when using FSDP + EP.
+# Initialize DeviceMesh for expert parallel (EP) modules when using HSDP + EP.
```
Comment (shjwudp) [inline megatron/core/distributed/fsdp/src/README.md]: `expert_device_mesh` is the device mesh of FSDP, I believe "FSDP + EP" is correct.

File: megatron/core/distributed/fsdp/src/README.md [hunk 4]
```diff
@@ -144,13 +144,13 @@ device_mesh[("dp_shard", "cp")]._flatten("dp_shard_cp")
...
 expert_device_mesh = torch.distributed.device_mesh.init_device_mesh(
     "cuda",
-    mesh_shape=(expt_dp_shard_size, expt_tp_size),
-    mesh_dim_names=("dp_shard", "tp"),
+    mesh_shape=(dp_outer_size, expt_dp_shard_size, expt_tp_size),
+    mesh_dim_names=("dp_outer", "dp_shard_cp", "tp"),
```
Comment (shjwudp) [inline megatron/core/distributed/fsdp/src/README.md:162]: This name `dp_shard_cp` is obfuscated; I suggest changing it to `dp_cp`.

File: megatron/core/distributed/fsdp/src/README.md [hunk 5]
```diff
@@ -144,13 +144,13 @@ device_mesh[("dp_shard", "cp")]._flatten("dp_shard_cp")
...
 )
-
+hsdp_expt_group = expert_device_mesh[("dp_outer", "dp_shard_cp")].get_group()
```
Comment (shjwudp) [inline megatron/core/distributed/fsdp/src/README.md]: Also note "dp_shard_cp" here.

File: megatron/core/distributed/fsdp/src/README.md [hunk 6]
```diff
@@ -260,6 +262,7 @@ optimizer.load_state_dict(ckpt_state_dict["optimizer"])
   - `tp_dim` is the name of the sub-mesh used for tensor parallelism (TP), which is required for `(FSDP, TP)`-strided sharding when using Megatron-LM or Torch-native `DTensor` TP.
     - For more information about tensor parallelism, refer to: [Megatron-LM: Training Multi-Billion Parameter Language Models Using Model Parallelism](https://arxiv.org/abs/1909.08053).
   - `hybrid_fsdp_group` is the `ProcessGroup` which contains all ranks in the flattened `dp_shard_dim` and `dp_outer_dim` sub-meshes utilized to specify the `(DP-Outer, DP-Shard)` sharded mesh coordinates for the weight and gradient buffers. Required for HSDP.
+  - `hybrid_fsdp_expt_group` is the equivalent `ProcessGroup` for Expert Parallel (EP) modules. It also contains all ranks from these flattened sub-meshes to establish the same `(DP-Outer, DP-Shard)` sharding coordinates and is required for combined HSDP + EP.
```
Comment (shjwudp) [inline megatron/core/distributed/fsdp/src/README.md]: We need the description to be direct and clear, for example:
`hybrid_fsdp_expt_group` defines the data-parallel communication group for expert parameters. It is required for HSDP.

File: megatron/core/distributed/fsdp/mcore_fsdp_adapter.py [hunk 7]
```diff
@@ -353,7 +381,20 @@ def _get_hsdp_tp_mesh(outer_fsdp_dp_group, dp_cp_group, tp_group):
         f"do not match the ranks in the Outer FSDP DP group {outer_fsdp_dp_group_ranks}."
     )
 
-    return mesh
+    # Exclude the expert parallel dimension
```
Comment (shjwudp) [inline megatron/core/distributed/fsdp/mcore_fsdp_adapter.py:388]: Export the device mesh containing the current rank.

File: megatron/core/distributed/fsdp/mcore_fsdp_adapter.py [hunk 8]
```diff
@@ -304,22 +329,24 @@ def sync_rng_states_across_tp_group(self):
...
 
-def _get_hsdp_tp_mesh(outer_fsdp_dp_group, dp_cp_group, tp_group):
+def _get_hsdp_tp_mesh(outer_fsdp_dp_group, dp_cp_group, tp_group, ep_size=1):
     assert HAVE_EINOPS, "einops is not installed. Please install it with `pip install einops`."
     world_size = dist.get_world_size()
 
     mesh = einops.rearrange(
         torch.arange(world_size),
-        "(outer_fsdp_dp fsdp tp) -> outer_fsdp_dp fsdp tp",
+        "(outer_fsdp_dp fsdp ep tp) -> ep outer_fsdp_dp fsdp tp",
```
Comment (shjwudp) [inline megatron/core/distributed/fsdp/mcore_fsdp_adapter.py:342]: Why change the relative position of EP here?

Comment (wplf) [reply megatron/core/distributed/fsdp/mcore_fsdp_adapter.py:342]: This maintains consistency with `_get_dp_tp_mesh()` in line 400 and makes it convenient to obtain the per-ep mesh at line 386.

This function returns the EP rank mesh which contains its only rank. 

For example, total ranks is `[0,1,2,3,4,5,6,7]`.
When ep_size==2,  total mesh will be
`[[0,1],
[2,3],
[4,5],
[6,7]]`

Move EP position outside will make mesh to be
`[[0,2,4,6], [1,3,5,7]]`, which makes it convenient to obtain the per-ep mesh at line 386.

File: megatron/core/distributed/fsdp/mcore_fsdp_adapter.py (hunk 7 above)
Comment (wplf) [reply megatron/core/distributed/fsdp/mcore_fsdp_adapter.py:388]: This is correct. `_get_dp_tp_mesh` follows the same convention.

Comment (shjwudp) [reply megatron/core/distributed/fsdp/mcore_fsdp_adapter.py:388]: I see. The description isn’t great, but `_get_dp_tp_mesh` and `_get_hsdp_tp_mesh` will be rewritten in a later [PR](https://github.com/NVIDIA/Megatron-LM/pull/2302/files#diff-8d3ed77aead9b21b07dfd6c4df4a3b0db421c1f29cf1d1ff52fcc72f8e2eccbfR300). We can aim for clearer and more concise descriptions in the future.

**Review by shjwudp — APPROVED** [review]

**Comment by shjwudp** [comment]
@wplf Please fix the linting error. You can use the following commands to identify any formatting issues:

```bash
export PATH=".venv/bin:$PATH"
export GITLAB_ENDPOINT=github.com
export CI_PROJECT_NAMESPACE=NVIDIA
export BASE_REF="dev"
export SKIP_DOCS=false
export CHECK_ONLY=true
bash tools/autoformat.sh
```

**Comment by wplf** [comment]
@shjwudp  Hi, jack. 
The linting error has been fixed. 
Please tell me if there is anything I can do.

**Comment by shjwudp** [comment]
Hi @NVIDIA/core-devtech  @NVIDIA/core-nemo , could you help review this PR? thanks.

**Comment by cspades** [comment]
Had nits on the main branch version of this PR, just documentation-related (use "dp_shard" and "dp_outer" terminology for public-facing documentation): https://github.com/NVIDIA/Megatron-LM/pull/2840

This PR can be merged immediately.

**Comment by wplf** [comment]
Hi, cory.
Do I need to rename `dp_cp` to `dp_shard_cp` in this branch?
Or do I need to rename it in both [PR-2800](https://github.com/NVIDIA/Megatron-LM/pull/2800) and [PR-2840](https://github.com/NVIDIA/Megatron-LM/pull/2800)?

I'm kind of confused.

**Comment by yanring** [comment]
Please add unit tests before the merging, thanks!

**Comment by wplf** [comment]
> Please add unit tests before the merging, thanks!

Hi, @yanring 
Jack has add unittest in the main branch, and I've merged it, and test it in [wandb](https://wandb.ai/megatron-core-moe-dev/jinliang-ds-hsdp-ep/panel/8hj8zz3wn?nw=nwuserjinliangl). 

- Main branch's HSDP + EP with outer dim no_shard is bugged.
- Main branch's HSDP + EP with outer shard optim will crash.

Our PR-2840 fix them and the outcome looks well.  cc @shjwudp 

Please let me know if there is anything else I can do.
image
